paper-api/src/main/java/io/papermc/paper/event/world/WorldDifficultyChangeEvent.java

```diff
@@ -0,0 +1,78 @@
+package io.papermc.paper.event.world;
+
+import org.bukkit.Difficulty;
+import org.bukkit.World;
+import org.bukkit.command.CommandSender;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.world.WorldEvent;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Called when a world's difficulty is changed, either by command or by api.
+ * <p>
+ * If the world is in {@link World#isHardcore() hardcore}
+ * the resulting difficulty will always be {@link Difficulty#HARD}
+ */
+@NullMarked
+public class WorldDifficultyChangeEvent extends WorldEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private final @Nullable CommandSender commandSender;
+    private Difficulty difficulty;
+    private boolean cancelled;
```

Comment:
Make this event immutable (only for reading changes). so no cancellation or setter for difficulty.